sensor/common/registry/registry_store.go

```diff
@@ -392,3 +392,221 @@ func (rs *Store) GetMatchingCentralRegistryIntegrations(imgName *storage.ImageNa
 
 	return regs
 }
+
+// UpsertSecret upserts a pull secret into the store.
+func (rs *Store) UpsertSecret(namespace, secretName string, dockerConfig config.DockerConfig, serviceAcctName string) {
+	if !features.SensorPullSecretsByName.Enabled() {
+		rs.upsertSecretByHost(namespace, secretName, dockerConfig, serviceAcctName)
+		return
+	}
+
+	rs.upsertSecretByName(namespace, secretName, dockerConfig, serviceAcctName)
+}
+
+func (rs *Store) upsertSecretByHost(namespace, secretName string, dockerConfig config.DockerConfig, serviceAcctName string) {
+	isGlobalPullSecret := namespace == openshiftConfigNamespace && secretName == openshiftConfigPullSecret
+	fromDefaultSA := serviceAcctName == defaultSA
+
+	for registryAddress, dce := range dockerConfig {
+		registryAddr := strings.TrimSpace(registryAddress)
+
+		if fromDefaultSA {
+			// We assume that registries found in the dockercfg secret managed by OCP for the default
+			// service account only references hostnames for the OCP internal registry.
+			rs.addClusterLocalRegistryHost(registryAddr)
+			if err := rs.upsertRegistry(namespace, registryAddr, dce); err != nil {
+				log.Errorf("Unable to upsert registry %q into store: %v", registryAddr, err)
+			}
+			continue
+		}
+
+		if env.DelegatedScanningDisabled.BooleanSetting() {
+			// If delegated scanning is disabled then we do not store additional secrets outside of those needed
+			// for scanning images from the OCP internal registry.
+			continue
+		}
+
+		if serviceAcctName != "" {
+			// Ignore secrets for service accounts other than default so that the
+			// default registry is not overwritten in the store.
+			continue
+		}
+
+		var err error
+		if isGlobalPullSecret {
+			err = rs.upsertGlobalRegistry(registryAddr, dce)
+		} else {
+			err = rs.upsertRegistry(namespace, registryAddr, dce)
+		}
+		if err != nil {
+			log.Errorf("unable to upsert registry %q into store: %v", registryAddr, err)
+		}
+	}
+}
+
+func (rs *Store) upsertSecretByName(namespace, secretName string, dockerConfig config.DockerConfig, serviceAcctName string) {
+	isGlobalPullSecret := namespace == openshiftConfigNamespace && secretName == openshiftConfigPullSecret
+	hasBoundServiceAccount := serviceAcctName != ""
+
+	// To avoid partial upserts - hold the lock until the entire secret upserted.
+	rs.storeMutux.Lock()
+	defer rs.storeMutux.Unlock()
+
+	for registryAddress, dce := range dockerConfig {
+		registryAddr := strings.TrimSpace(registryAddress)
+
+		if hasBoundServiceAccount {
+			// We assume that registries found in dockercfg secrets managed by OCP only
+			// reference hostnames for the OCP internal registry.
```

Comment:
how does this function know this is an OCP-managed secret?

Reply:
It is OCP managed if the secret is bound to a service account (same as the `default` secrets) - relates to your previous comment.  The `default` SA is not special in that regards, every service account will have secrets 'bound' to it that contain credentials for accessing the OCP internal registry.